packages/runtime-common/virtual-network.ts

```diff
@@ -86,3 +206,63 @@ function isUrlLike(moduleIdentifier: string): boolean {
     moduleIdentifier.startsWith('https://')
   );
 }
+
+async function getContentOfReadableStream(
+  requestBody: ReadableStream<Uint8Array> | null,
+): Promise<Uint8Array | null> {
+  if (requestBody) {
+    let isPending = true;
+    let arrayLegnth = 0;
+    let unit8Arrays = [];
+    let reader = requestBody.getReader();
+    do {
+      let readableResults = await reader.read();
+
+      if (readableResults.value) {
+        arrayLegnth += readableResults.value.length;
+        unit8Arrays.push(readableResults.value);
+      }
+
+      isPending = !readableResults.done;
+    } while (isPending);
+    let mergedArray = new Uint8Array(arrayLegnth);
+    unit8Arrays.forEach((array) => mergedArray.set(array));
+    return mergedArray;
+  }
+  return null;
+}
+
+async function buildRequest(url: string, originalRequest: Request) {
+  if (url === originalRequest.url) {
+    return originalRequest;
+  }
+
+  // To reach the goal of creating a new Request but with a different url it is
+  // usually enough to create a new Request object with the new url and the same
+  // properties as the original request, but there are issues when the body is
+  // a ReadableStream - browser reports the following error:
+  // "TypeError: Failed to construct 'Request': The `duplex` member must be
+  // specified for a request with a streaming body." Even adding the `duplex`
+  // property will not fix the issue - the browser request being made to
+  // our local server then expects HTTP/2 connection which is currently not
+  // supported in our local server. To avoid all these issues, we resort to
+  // reading the body of the original request and creating a new Request with
+  // the new url and the body as a Uint8Array.
+
+  let body = null;
+  if (originalRequest.body) {
+    body = await getContentOfReadableStream(originalRequest.clone().body);
```

Comment:
How does this work with SSE, or does it need to? My understanding of SSE is that the body never is "finished".

Reply:
Native browser API is used to subscribe to an SSE event--we don't go thru a fetch (https://github.com/cardstack/boxel/blob/main/packages/host/app/services/message-service.ts#L19). Given that I wonder if you still need to unresolve that URL that you pass to the `new EventSource(url)` that we use to subscribe to SSE events. Right now only the base realm really has a URL that is different between the resolved and unresolved forms--so if we are not subscribing to the base realm SSE (and I don't think we ever do), it's not something that you'd be able to see (so like a hidden bug).

Reply:
That's right, SSE uses an `EventSource` - https://github.com/cardstack/boxel/pull/1126/files#diff-cfc037186d9054c397cd997d5445153953255681ca1d5e29b02d4278345763b3R20

Regarding subscribing to base realm using SSE - I observed that we actually do subscribe to the base realm SSE, for example if you visit `https://cardstack.com/base/boolean.gts` then the file resource will subscribe to the base realm.

Reply:
Yes, the change is reflected in the host code editor if I change the boolean file locally. It will correctly connect to the resolved url via event source. So I think this is all good?